distributed/worker.py

```diff
@@ -3044,6 +3046,8 @@ async def gather_dep(
                 for d in has_what:
                     ts = self.tasks[d]
                     ts.who_has.remove(worker)
+                    if ts.state == "released":
+                        continue
                     if not ts.who_has:
                         recommendations[ts] = "missing"
```

Comment:
We shouldn't attempt to transition to missing if the states are not in fetch or flight. E.g. a task can be in memory but it will not show up in who_has.
Wouldn't be surprised to see other states also possible here.
```suggestion
                    ts.who_has.remove(worker)
                    if not ts.who_has and ts.state in {"fetch", "flight"}:
                        recommendations[ts] = "missing"
```

Overall we should break up this response handling. I consider it a big problem that we're executing the task transitioning and recommendation calculation as part of a finally clause.

Reply:
Yeah, `missing` is only a relevant state for data-fetch tasks, right? If we have a runspec for a task (meaning we're supposed to execute it), it being `missing` is meaningless.

I wonder how much worker state complexity and confusion is due to the fact that we use the same system for both tasks to run and keys to fetch from other workers...

Reply:
Implemented the suggestion here